File: frontend/appflowy_tauri/src/appflowy_app/stores/effects/database/row/row_cache.ts
import {
  RowPB,
  InsertedRowPB,
  UpdatedRowPB,
  RowIdPB,
  OptionalRowPB,
  RowsChangePB,
  RowsVisibilityChangePB,
  ReorderSingleRowPB,
} from '@/services/backend';
import { ChangeNotifier } from '$app/utils/change_notifier';
import { FieldInfo } from '../field/field_controller';
import { CellCache, CellCacheKey } from '../cell/cell_cache';
import { CellIdentifier } from '../cell/cell_bd_svc';
import { DatabaseEventGetRow } from '@/services/backend/events/flowy-database2';
import { None, Option, Some } from 'ts-results';
import { Log } from '$app/utils/log';

export type CellByFieldId = Map<string, CellIdentifier>;

export class RowCache {
  private readonly rowList: RowList;
  private readonly cellCache: CellCache;
  private readonly notifier: RowChangeNotifier;

  constructor(public readonly viewId: string, private readonly getFieldInfos: () => readonly FieldInfo[]) {
    this.rowList = new RowList();
    this.cellCache = new CellCache(viewId);
    this.notifier = new RowChangeNotifier();
  }

  get rows(): readonly RowInfo[] {
    return this.rowList.rows;
  }

  getCellCache = () => {
    return this.cellCache;
  };

  loadCells = async (rowId: string): Promise<CellByFieldId> => {
    const opRow = this.rowList.getRow(rowId);
    if (opRow.some) {
      return this._toCellMap(opRow.val.row.id, this.getFieldInfos());
    } else {
      const rowResult = await this._loadRow(rowId);
      if (rowResult.ok) {
        this._refreshRow(rowResult.val);
        return this._toCellMap(rowId, this.getFieldInfos());
      } else {
        Log.error(rowResult.val);
        return new Map();
      }
    }
  };

  subscribe = (callbacks: {
    onRowsChanged: (reason: RowChangedReason, cellMap?: Map<string, CellIdentifier>) => void;
  }) => {
    return this.notifier.observer?.subscribe((change) => {
      if (change.rowId !== undefined) {
        callbacks.onRowsChanged(change.reason, this._toCellMap(change.rowId, this.getFieldInfos()));
      } else {
        callbacks.onRowsChanged(change.reason);
      }
    });
  };

  onFieldUpdated = (fieldInfo: FieldInfo) => {
    // Remove the cell data if the corresponding field was changed
    this.cellCache.removeWithFieldId(fieldInfo.field.id);
  };

  onNumberOfFieldsUpdated = (fieldInfos: readonly FieldInfo[]) => {
    this.rowList.setFieldInfos(fieldInfos);
    this.notifier.withChange(RowChangedReason.FieldDidChanged);
  };

  initializeRows = (rows: RowPB[]) => {
    rows.forEach((rowPB) => {
      this.rowList.push(this._toRowInfo(rowPB));
    });
    this.notifier.withChange(RowChangedReason.ReorderRows);
  };

  applyRowsChanged = (changeset: RowsChangePB) => {
    this._deleteRows(changeset.deleted_rows);
    this._insertRows(changeset.inserted_rows);
    this._updateRows(changeset.updated_rows);
  };

  applyRowsVisibility = (changeset: RowsVisibilityChangePB) => {
    this._hideRows(changeset.invisible_rows);
    this._displayRows(changeset.visible_rows);
  };

  applyReorderRows = (rowIds: string[]) => {
    this.rowList.reorderByRowIds(rowIds);
    this.notifier.withChange(RowChangedReason.ReorderRows);
  };

  applyReorderSingleRow = (reorderRow: ReorderSingleRowPB) => {
    const rowInfo = this.rowList.getRow(reorderRow.row_id);
    if (rowInfo !== undefined) {
      this.rowList.move({ rowId: reorderRow.row_id, fromIndex: reorderRow.old_index, toIndex: reorderRow.new_index });
      this.notifier.withChange(RowChangedReason.ReorderSingleRow, reorderRow.row_id);
    }
  };

  private _refreshRow = (opRow: OptionalRowPB) => {
    if (!opRow.has_row) {
      return;
    }
    const updatedRow = opRow.row;
    const option = this.rowList.getRowWithIndex(updatedRow.id);
    if (option.some) {
      const { rowInfo, index } = option.val;
      this.rowList.remove(rowInfo.row.id);
      this.rowList.insert(index, rowInfo.copyWith({ row: updatedRow }));
    } else {
      const newRowInfo = new RowInfo(this.viewId, this.getFieldInfos(), updatedRow);
      this.rowList.push(newRowInfo);
    }
  };

  private _loadRow = (rowId: string) => {
    const payload = RowIdPB.fromObject({ view_id: this.viewId, row_id: rowId });
    return DatabaseEventGetRow(payload);
  };

  private _deleteRows = (rowIds: string[]) => {
    rowIds.forEach((rowId) => {
      const deletedRow = this.rowList.remove(rowId);
      if (deletedRow !== undefined) {
        this.notifier.withChange(RowChangedReason.Delete, deletedRow.rowInfo.row.id);
      }
    });
  };

  private _insertRows = (rows: InsertedRowPB[]) => {
    rows.forEach((insertedRow) => {
      const rowInfo = this._toRowInfo(insertedRow.row);
      const insertedIndex = this.rowList.insert(insertedRow.index, rowInfo);
      if (insertedIndex !== undefined) {
        this.notifier.withChange(RowChangedReason.Insert, insertedIndex.rowId);
      }
    });
  };

  private _updateRows = (updatedRows: UpdatedRowPB[]) => {
    if (updatedRows.length === 0) {
      return;
    }

    const rowInfos: RowInfo[] = [];
    updatedRows.forEach((updatedRow) => {
      updatedRow.field_ids.forEach((fieldId) => {
        const key = new CellCacheKey(fieldId, updatedRow.row.id);
        this.cellCache.remove(key);
      });

      rowInfos.push(this._toRowInfo(updatedRow.row));
    });

    const updatedIndexs = this.rowList.insertRows(rowInfos);
    updatedIndexs.forEach((row) => {
      this.notifier.withChange(RowChangedReason.Update, row.rowId);
    });
  };

  private _hideRows = (rowIds: string[]) => {
    rowIds.forEach((rowId) => {
      const deletedRow = this.rowList.remove(rowId);
      if (deletedRow !== undefined) {
        this.notifier.withChange(RowChangedReason.Delete, deletedRow.rowInfo.row.id);
      }
    });
  };

  private _displayRows = (insertedRows: InsertedRowPB[]) => {
    insertedRows.forEach((insertedRow) => {
      const insertedIndex = this.rowList.insert(insertedRow.index, this._toRowInfo(insertedRow.row));

      if (insertedIndex !== undefined) {
        this.notifier.withChange(RowChangedReason.Insert, insertedIndex.rowId);
      }
    });
  };

  dispose = async () => {
    this.notifier.dispose();
  };

  private _toRowInfo = (rowPB: RowPB) => {
    return new RowInfo(this.viewId, this.getFieldInfos(), rowPB);
  };

  private _toCellMap = (rowId: string, fieldInfos: readonly FieldInfo[]): CellByFieldId => {
    const cellIdentifierByFieldId: Map<string, CellIdentifier> = new Map();

    fieldInfos.forEach((fieldInfo) => {
      const identifier = new CellIdentifier(this.viewId, rowId, fieldInfo.field.id, fieldInfo.field.field_type);
      cellIdentifierByFieldId.set(fieldInfo.field.id, identifier);
    });

    return cellIdentifierByFieldId;
  };
}

class RowList {
  _rowInfos: RowInfo[] = [];
  _rowInfoByRowId: Map<string, RowInfo> = new Map();

  get rows(): readonly RowInfo[] {
    return this._rowInfos;
  }

  getRow = (rowId: string): Option<RowInfo> => {
    const rowInfo = this._rowInfoByRowId.get(rowId);
    if (rowInfo === undefined) {
      return None;
    } else {
      return Some(rowInfo);
    }
  };
  getRowWithIndex = (rowId: string): Option<{ rowInfo: RowInfo; index: number }> => {
    const rowInfo = this._rowInfoByRowId.get(rowId);
    if (rowInfo !== undefined) {
      const index = this._rowInfos.indexOf(rowInfo, 0);
      return Some({ rowInfo: rowInfo, index: index });
    }
    return None;
  };

  indexOfRow = (rowId: string): number => {
    const rowInfo = this._rowInfoByRowId.get(rowId);
    if (rowInfo !== undefined) {
      return this._rowInfos.indexOf(rowInfo, 0);
    }
    return -1;
  };

  push = (rowInfo: RowInfo) => {
    const index = this.indexOfRow(rowInfo.row.id);
    if (index !== -1) {
      this._rowInfos.splice(index, 1, rowInfo);
    } else {
      this._rowInfos.push(rowInfo);
    }

    this._rowInfoByRowId.set(rowInfo.row.id, rowInfo);
  };

  remove = (rowId: string): DeletedRow | undefined => {
    const result = this.getRowWithIndex(rowId);
    if (result.some) {
      const { rowInfo, index } = result.val;
      this._rowInfoByRowId.delete(rowInfo.row.id);
      this._rowInfos.splice(index, 1);
      return new DeletedRow(index, rowInfo);
    } else {
      return undefined;
    }
  };

  insert = (insertIndex: number, newRowInfo: RowInfo): InsertedRow | undefined => {
    const rowId = newRowInfo.row.id;
    // Calibrate where to insert
    let insertedIndex = insertIndex;
    if (this._rowInfos.length <= insertedIndex) {
      insertedIndex = this._rowInfos.length;
    }
    const result = this.getRowWithIndex(rowId);

    if (result.some) {
      const { index } = result.val;
      // remove the old row info
      this._rowInfos.splice(index, 1);
      // insert the new row info to the insertedIndex
      this._rowInfos.splice(insertedIndex, 0, newRowInfo);
      this._rowInfoByRowId.set(rowId, newRowInfo);
      return undefined;
    } else {
      this._rowInfos.splice(insertedIndex, 0, newRowInfo);
      this._rowInfoByRowId.set(rowId, newRowInfo);
      return new InsertedRow(insertedIndex, rowId);
    }
  };

  insertRows = (rowInfos: RowInfo[]) => {
    const map = new Map<string, InsertedRow>();
    rowInfos.forEach((rowInfo) => {
      const index = this.indexOfRow(rowInfo.row.id);
      if (index !== -1) {
        this._rowInfos.splice(index, 1, rowInfo);
        this._rowInfoByRowId.set(rowInfo.row.id, rowInfo);

        map.set(rowInfo.row.id, new InsertedRow(index, rowInfo.row.id));
      }
    });
    return map;
  };

  move = (params: { rowId: string; fromIndex: number; toIndex: number }) => {
    const currentIndex = this.indexOfRow(params.rowId);
    if (currentIndex !== -1 && currentIndex !== params.toIndex) {
      const rowInfo = this.remove(params.rowId)?.rowInfo;
      if (rowInfo !== undefined) {
        this.insert(params.toIndex, rowInfo);
      }
    }
  };

  reorderByRowIds = (rowIds: string[]) => {
    // remove all the elements
    this._rowInfos = [];
    rowIds.forEach((rowId) => {
      const rowInfo = this._rowInfoByRowId.get(rowId);
      if (rowInfo !== undefined) {
        this._rowInfos.push(rowInfo);
      }
    });
  };

  includes = (rowId: string): boolean => {
    return this._rowInfoByRowId.has(rowId);
  };

  setFieldInfos = (fieldInfos: readonly FieldInfo[]) => {
    const newRowInfos: RowInfo[] = [];
    this._rowInfos.forEach((rowInfo) => {
      newRowInfos.push(rowInfo.copyWith({ fieldInfos: fieldInfos }));
    });
    this._rowInfos = newRowInfos;
  };
}

export class RowInfo {
  constructor(
    public readonly viewId: string,
    public readonly fieldInfos: readonly FieldInfo[],
    public readonly row: RowPB
  ) {}

  copyWith = (params: { row?: RowPB; fieldInfos?: readonly FieldInfo[] }) => {
    return new RowInfo(this.viewId, params.fieldInfos || this.fieldInfos, params.row || this.row);
  };
}

export class DeletedRow {
  constructor(public readonly index: number, public readonly rowInfo: RowInfo) {}
}

export class InsertedRow {
  constructor(public readonly index: number, public readonly rowId: string) {}
}

export class RowChanged {
  constructor(public readonly reason: RowChangedReason, public readonly rowId?: string) {}
}

// eslint-disable-next-line no-shadow
export enum RowChangedReason {
  Insert,
  Delete,
  Update,
  Initial,
  FieldDidChanged,
  ReorderRows,
  ReorderSingleRow,
}

export class RowChangeNotifier extends ChangeNotifier<RowChanged> {
  _currentChanged = new RowChanged(RowChangedReason.Initial);

  withChange = (reason: RowChangedReason, rowId?: string) => {
    const newChange = new RowChanged(reason, rowId);
    if (this._currentChanged !== newChange) {
      this._currentChanged = newChange;
      this.notify(this._currentChanged);
    }
  };

  dispose = () => {
    this.unsubscribe();
  };
}

File: frontend/appflowy_flutter/lib/plugins/database_view/application/row/row_cache.dart
import 'dart:collection';
import 'package:appflowy/plugins/database_view/application/field/field_info.dart';
import 'package:appflowy_backend/dispatch/dispatch.dart';
import 'package:appflowy_backend/log.dart';
import 'package:appflowy_backend/protobuf/flowy-database2/protobuf.dart';
import 'package:flutter/foundation.dart';
import 'package:freezed_annotation/freezed_annotation.dart';

import '../cell/cell_service.dart';
import 'row_list.dart';
import 'row_service.dart';
part 'row_cache.freezed.dart';

typedef RowUpdateCallback = void Function();

/// A delegate that provides the fields of the row.
abstract class RowFieldsDelegate {
  UnmodifiableListView<FieldInfo> get fieldInfos;
  void onFieldsChanged(void Function(List<FieldInfo>) callback);
}

abstract mixin class RowLifeCycle {
  void onRowDisposed();
}

/// Read https://appflowy.gitbook.io/docs/essential-documentation/contribute-to-appflowy/architecture/frontend/grid for more information.

class RowCache {
  final String viewId;
  final RowList _rowList = RowList();
  final CellMemCache _cellMemCache;
  final RowLifeCycle _rowLifeCycle;
  final RowFieldsDelegate _fieldDelegate;
  final RowChangesetNotifier _changedNotifier;

  /// Returns a unmodifiable list of RowInfo
  UnmodifiableListView<RowInfo> get rowInfos {
    final visibleRows = [..._rowList.rows];
    return UnmodifiableListView(visibleRows);
  }

  /// Returns a unmodifiable map of RowInfo
  UnmodifiableMapView<RowId, RowInfo> get rowByRowId {
    return UnmodifiableMapView(_rowList.rowInfoByRowId);
  }

  CellMemCache get cellCache => _cellMemCache;
  ChangedReason get changeReason => _changedNotifier.reason;

  RowCache({
    required this.viewId,
    required RowFieldsDelegate fieldsDelegate,
    required RowLifeCycle rowLifeCycle,
  })  : _cellMemCache = CellMemCache(viewId: viewId),
        _changedNotifier = RowChangesetNotifier(),
        _rowLifeCycle = rowLifeCycle,
        _fieldDelegate = fieldsDelegate {
    // Listen on the changed of the fields. If the fields changed, we need to
    // clear the cell cache with the given field id.
    fieldsDelegate.onFieldsChanged((fieldInfos) {
      for (final fieldInfo in fieldInfos) {
        _cellMemCache.removeCellWithFieldId(fieldInfo.id);
      }
      _changedNotifier.receive(const ChangedReason.fieldDidChange());
    });
  }

  RowInfo? getRow(RowId rowId) {
    return _rowList.get(rowId);
  }

  void setInitialRows(List<RowMetaPB> rows) {
    for (final row in rows) {
      final rowInfo = buildGridRow(row);
      _rowList.add(rowInfo);
    }
    _changedNotifier.receive(const ChangedReason.setInitialRows());
  }

  Future<void> dispose() async {
    _rowLifeCycle.onRowDisposed();
    _changedNotifier.dispose();
    await _cellMemCache.dispose();
  }

  void applyRowsChanged(RowsChangePB changeset) {
    _deleteRows(changeset.deletedRows);
    _insertRows(changeset.insertedRows);
    _updateRows(changeset.updatedRows);
  }

  void applyRowsVisibility(RowsVisibilityChangePB changeset) {
    _hideRows(changeset.invisibleRows);
    _showRows(changeset.visibleRows);
  }

  void reorderAllRows(List<String> rowIds) {
    _rowList.reorderWithRowIds(rowIds);
    _changedNotifier.receive(const ChangedReason.reorderRows());
  }

  void reorderSingleRow(ReorderSingleRowPB reorderRow) {
    final rowInfo = _rowList.get(reorderRow.rowId);
    if (rowInfo != null) {
      _rowList.moveRow(
        reorderRow.rowId,
        reorderRow.oldIndex,
        reorderRow.newIndex,
      );
      _changedNotifier.receive(
        ChangedReason.reorderSingleRow(
          reorderRow,
          rowInfo,
        ),
      );
    }
  }

  void _deleteRows(List<RowId> deletedRowIds) {
    for (final rowId in deletedRowIds) {
      final deletedRow = _rowList.remove(rowId);
      if (deletedRow != null) {
        _changedNotifier.receive(ChangedReason.delete(deletedRow));
      }
    }
  }

  void _insertRows(List<InsertedRowPB> insertRows) {
    for (final insertedRow in insertRows) {
      final insertedIndex =
          _rowList.insert(insertedRow.index, buildGridRow(insertedRow.rowMeta));
      if (insertedIndex != null) {
        _changedNotifier.receive(ChangedReason.insert(insertedIndex));
      }
    }
  }

  void _updateRows(List<UpdatedRowPB> updatedRows) {
    if (updatedRows.isEmpty) return;
    final List<RowMetaPB> updatedList = [];
    for (final updatedRow in updatedRows) {
      for (final fieldId in updatedRow.fieldIds) {
        final key = CellCacheKey(
          fieldId: fieldId,
          rowId: updatedRow.rowId,
        );
        _cellMemCache.remove(key);
      }
      if (updatedRow.hasRowMeta()) {
        updatedList.add(updatedRow.rowMeta);
      }
    }

    final updatedIndexs =
        _rowList.updateRows(updatedList, (rowId) => buildGridRow(rowId));

    if (updatedIndexs.isNotEmpty) {
      _changedNotifier.receive(ChangedReason.update(updatedIndexs));
    }
  }

  void _hideRows(List<RowId> invisibleRows) {
    for (final rowId in invisibleRows) {
      final deletedRow = _rowList.remove(rowId);
      if (deletedRow != null) {
        _changedNotifier.receive(ChangedReason.delete(deletedRow));
      }
    }
  }

  void _showRows(List<InsertedRowPB> visibleRows) {
    for (final insertedRow in visibleRows) {
      final insertedIndex =
          _rowList.insert(insertedRow.index, buildGridRow(insertedRow.rowMeta));
      if (insertedIndex != null) {
        _changedNotifier.receive(ChangedReason.insert(insertedIndex));
      }
    }
  }

  void onRowsChanged(void Function(ChangedReason) onRowChanged) {
    _changedNotifier.addListener(() {
      onRowChanged(_changedNotifier.reason);
    });
  }

  RowUpdateCallback addListener({
    required RowId rowId,
    void Function(CellContextByFieldId, ChangedReason)? onRowChanged,
  }) {
    listenerHandler() async {
      if (onRowChanged != null) {
        final rowInfo = _rowList.get(rowId);
        if (rowInfo != null) {
          final cellDataMap = _makeCells(rowInfo.rowMeta);
          onRowChanged(cellDataMap, _changedNotifier.reason);
        }
      }
    }

    _changedNotifier.addListener(listenerHandler);
    return listenerHandler;
  }

  void removeRowListener(VoidCallback callback) {
    _changedNotifier.removeListener(callback);
  }

  CellContextByFieldId loadCells(RowMetaPB rowMeta) {
    final rowInfo = _rowList.get(rowMeta.id);
    if (rowInfo == null) {
      _loadRow(rowMeta.id);
    }
    return _makeCells(rowMeta);
  }

  Future<void> _loadRow(RowId rowId) async {
    final payload = RowIdPB.create()
      ..viewId = viewId
      ..rowId = rowId;

    final result = await DatabaseEventGetRowMeta(payload).send();
    result.fold(
      (rowMetaPB) {
        final rowInfo = _rowList.get(rowMetaPB.id);
        final rowIndex = _rowList.indexOfRow(rowMetaPB.id);
        if (rowInfo != null && rowIndex != null) {
          final updatedRowInfo = rowInfo.copyWith(rowMeta: rowMetaPB);
          _rowList.remove(rowMetaPB.id);
          _rowList.insert(rowIndex, updatedRowInfo);

          final UpdatedIndexMap updatedIndexs = UpdatedIndexMap();
          updatedIndexs[rowMetaPB.id] = UpdatedIndex(
            index: rowIndex,
            rowId: rowMetaPB.id,
          );

          _changedNotifier.receive(ChangedReason.update(updatedIndexs));
        }
      },
      (err) => Log.error(err),
    );
  }

  CellContextByFieldId _makeCells(RowMetaPB rowMeta) {
    // ignore: prefer_collection_literals
    final cellContextMap = CellContextByFieldId();
    for (final fieldInfo in _fieldDelegate.fieldInfos) {
      cellContextMap[fieldInfo.id] = DatabaseCellContext(
        rowMeta: rowMeta,
        viewId: viewId,
        fieldInfo: fieldInfo,
      );
    }
    return cellContextMap;
  }

  RowInfo buildGridRow(RowMetaPB rowMetaPB) {
    return RowInfo(
      viewId: viewId,
      fields: _fieldDelegate.fieldInfos,
      rowId: rowMetaPB.id,
      rowMeta: rowMetaPB,
    );
  }
}

class RowChangesetNotifier extends ChangeNotifier {
  ChangedReason reason = const InitialListState();

  RowChangesetNotifier();

  void receive(ChangedReason newReason) {
    reason = newReason;
    reason.map(
      insert: (_) => notifyListeners(),
      delete: (_) => notifyListeners(),
      update: (_) => notifyListeners(),
      fieldDidChange: (_) => notifyListeners(),
      initial: (_) {},
      reorderRows: (_) => notifyListeners(),
      reorderSingleRow: (_) => notifyListeners(),
      setInitialRows: (_) => notifyListeners(),
    );
  }
}

@unfreezed
class RowInfo with _$RowInfo {
  factory RowInfo({
    required String rowId,
    required String viewId,
    required UnmodifiableListView<FieldInfo> fields,
    required RowMetaPB rowMeta,
  }) = _RowInfo;
}

typedef InsertedIndexs = List<InsertedIndex>;
typedef DeletedIndexs = List<DeletedIndex>;
// key: id of the row
// value: UpdatedIndex
typedef UpdatedIndexMap = LinkedHashMap<RowId, UpdatedIndex>;

@freezed
class ChangedReason with _$ChangedReason {
  const factory ChangedReason.insert(InsertedIndex item) = _Insert;
  const factory ChangedReason.delete(DeletedIndex item) = _Delete;
  const factory ChangedReason.update(UpdatedIndexMap indexs) = _Update;
  const factory ChangedReason.fieldDidChange() = _FieldDidChange;
  const factory ChangedReason.initial() = InitialListState;
  const factory ChangedReason.reorderRows() = _ReorderRows;
  const factory ChangedReason.reorderSingleRow(
    ReorderSingleRowPB reorderRow,
    RowInfo rowInfo,
  ) = _ReorderSingleRow;
  const factory ChangedReason.setInitialRows() = _SetInitialRows;
}

class InsertedIndex {
  final int index;
  final RowId rowId;
  InsertedIndex({
    required this.index,
    required this.rowId,
  });
}

class DeletedIndex {
  final int index;
  final RowInfo rowInfo;
  DeletedIndex({
    required this.index,
    required this.rowInfo,
  });
}

class UpdatedIndex {
  final int index;
  final RowId rowId;
  UpdatedIndex({
    required this.index,
    required this.rowId,
  });
}
